Main svc events

elbonreedZ opened the pull request.

elbonreedZ committed:
Начальная стадия Events

elbonreedZ committed:
Реализовала Event

no-vi-ka commented on main-service/src/main/java/ru/practicum/event/dto/EventAdminParam.java:
```diff
@@ -1,18 +1,14 @@
 
-@Getter
-@Setter
+@Data
```
Почему ты заменила геттер и сеттер на аннотацию @Data? В прошлый раз ревьюер писал, что две отдельных аннотации - это более предпочтительный вариант.

elbonreedZ replied:
ну да, но так проще, потому что это не основная сущность а дтошки. И метод ту стринг еще нужен иначе логи не информативные

elbonreedZ replied:
Да, я еще посмотрела щас и это нужно только для entity, работающей с jpa

EyesHead commented on main-service/src/main/java/ru/practicum/event/service/EventServiceImpl.java:
```diff
@@ -0,0 +1,285 @@
+
+    @Override
+    public EventFullDto getByIdPublic(long eventId, StatDto statDto) {
+        Event event = findById(eventId);
```
Наличие подобных вызовов не следует использовать, так как метод findById - фактически является реализацией интерфейса EventService. Было обсуждение на эту тему, точно помню. Нарушает первый и второй принципы SOLID. В таких местах во всем коде следует писать целиком код, тем более что он станет длинее всего на 1 строчку)
```java
        Event event = eventRepository.findById(eventId).orElseThrow(() ->
                new NotFoundException("Событие с id: " + eventId + " не существует"));
```
Касается точно так же и приватных методов, которые сохранают одну строчку, зато не соответствуют Single Responsobility Principle (так как помимо возврата объекта они еще и обрабатывают исключение - если объект не найден)

В общем все вызовы репозиториев надо делать сразу в реализации метода интерфейса

elbonreedZ replied:
С 1 я частично согласна, я его публичным сделала чтобы его можно было например в твоих реквестах использовать, можно заменить на приватный. Но со вторым совершенно нет. Есть еще принцип DRY как никак. От того что я сразу обрабатываю исключение ничего критичного нет, логика поиска всегда такая. Ты говоришь, что станет больше на одну строчку, а на самом деле вызываю раз 5 этот метод, и получается уже на 5 строчек, а когда еще несколько сущностей добавляется, то еще больше

EyesHead commented on main-service/src/main/java/ru/practicum/event/service/EventServiceImpl.java:
```diff
@@ -0,0 +1,285 @@
+
+        List<EventShortDto> eventShorts = eventMapper.toEventShortDtoList(events);
+
+        return makeEventsWithViews(eventShorts);
```
С названиями приватных методов стоит быть осторожнее. Конкретно этот по названию должен СОЗДАТЬ ивент с просмотрами, фактически он ПРИМЕНЯЕТ к каждому ивенту из передаваемого списка просмотры.
По названию applyViewsToEvents или getEventsWithViews было бы более ожидаемо, что делает метод и что он вернет

elbonreedZ replied:
Он не создает а делает. Ну в принципе могу изменить, но не считаю, что это стоит

EyesHead commented on main-service/src/main/java/ru/practicum/event/service/EventServiceImpl.java:
```diff
@@ -0,0 +1,285 @@
+
+    @Override
+    public List<EventShortDto> getAllByUser(long userId, Pageable pageable) {
+        return eventMapper.toEventShortDtoList(eventRepository.findAllByInitiator_Id(userId, pageable));
```
Не стоит в качестве аргумента передавать функцию в java, очень сокращает читабельность кода. Лучше в таких местах всегда использовать локальные переменные.

Объявлять их нужно как можно ближе к их вызову (не конкретно про этот случай, а в общем)

elbonreedZ replied:
Субъективно. Просто return в одну строку недели создавать переменную для одного применения. Но в принципе могу изменить.

EyesHead commented on main-service/src/main/java/ru/practicum/event/service/EventServiceImpl.java:
```diff
@@ -0,0 +1,285 @@
+        if (stateAction != null) {
+            switch (stateAction) {
+                case PUBLISH_EVENT -> {
+                    if (!event.getState().equals(EventState.PENDING)) {
```
подобную логику стоит вынести в отдельные private методы, по типу handlePublishEvent и handleRejectEvent.

elbonreedZ replied:
согласна

EyesHead commented on main-service/src/main/java/ru/practicum/event/service/EventServiceImpl.java:
```diff
@@ -0,0 +1,285 @@
+                    }
+                }
+                case REJECT_EVENT -> {
+                    if (event.getState().equals(EventState.PENDING)) {
```
В подобных участках старайся всегда использовать Fail-fast подход. Если коротко, то fail-fast - это подход, при котором код как можно раньше проверяет условия, которые могут привести к ошибке, и сразу же завершает выполнение с соответствующим исключением или ошибкой. Советую с ним ознакомится, так как используется вообще всегда
``` java
   if (!EventState.PENDING.equals(event.getState())) {
        throw new ConditionsNotMetException("Нельзя отклонить опубликованное событие");
    }
    event.setState(EventState.CANCELED);
   ```

elbonreedZ replied:
Я так и делаю. И мой код даже в данном конкретном случае соответствует. Только ответвление есть у условия. То исключение бросается так же сразу. В общем вопрос только в предпочитаемом стиле

Commit:
Добавлен класс для параметров - UserAdminParam
Произведен рефакторинг сервиса (вместо параметров метода используется объект класса UserAdminParam)

elbonreedZ committed:
Поправила недочеты

elbonreedZ committed:
Merge remote-tracking branch 'origin/main_svc_events' into main_svc_events
